File: scala/21.scala
import scala.math.sqrt

def divisors(n: Int) = {
  val d = (2 to sqrt(n).toInt).filter(n % _ == 0)
  Set() ++ d ++ d.map(n / _) ++ Set(1)
}

def sigma(n: Int) = divisors(n).reduceLeft(_ + _)

val result = (2 to 10000).map(n => (n, sigma(n)))
    .filter(t => (t._1 > t._2 && sigma(t._2) == t._1))
    .map(t => t._1 + t._2)
    .reduceLeft(_ + _)

println(result)
assert(result == 31626)

File: solns/haskell/21.hs
import Data.List (nub)

divisors n = let sqn = floor . sqrt . fromIntegral $ n
                 a = filter (\q -> mod n q == 0) [2..sqn]
                 b = map (div n) a
             in nub ([1] ++ a ++ b)

sigma = sum . divisors

range = [2..10000]
result = let tuples = zip range (map sigma range)
             amicables = filter (\(a,b) -> a > b && a == sigma b) tuples
         in sum $ map (uncurry (+)) amicables

main = print result
